palace/utils/iodata.hpp

```diff
@@ -47,18 +48,34 @@ class IoData
   // Return the mesh scaling factor in units model.L0 x [m] for mesh IO.
   double GetLengthScale() const { return Lc / model.L0; }
 
-  // Redimensionalize values for output.
+  // Redimensionalize values for output. Outputs which depend on the fields assume a
+  // characteristic reference magnetic field strength Hc such that Pc = 1 W, where Pc is the
+  // characteristic reference power.
   enum class ValueType
   {
-    TIME,         // [ns]
-    FREQUENCY,    // [GHz]
-    LENGTH,       // [m]
-    IMPEDANCE,    // [Ω]
-    INDUCTANCE,   // [H]
-    CAPACITANCE,  // [F]
-    CONDUCTIVITY  // [S/m]
+    TIME,          // [ns]
+    FREQUENCY,     // [GHz]
+    LENGTH,        // [m]
+    IMPEDANCE,     // [Ω]
+    INDUCTANCE,    // [H] = [Ωs]
+    CAPACITANCE,   // [F] = [s/Ω]
+    CONDUCTIVITY,  // [S/m]
+    VOLTAGE,       // [V]
+    CURRENT,       // [A]
+    POWER,         // [W]
+    ENERGY,        // [J]
+    FIELD_E,       // [V/m]
+    FIELD_D,       // [C/m²] = [A⋅s/m²]
+    FIELD_H,       // [A/m]
+    FIELD_B        // [Wb/m²] = [V⋅s/m²]
   };
-  double DimensionalizeValue(ValueType type, double v) const;
+  template <typename T>
+  T DimensionalizeValue(ValueType type, T v) const;
+  template <typename T>
+  std::complex<T> DimensionalizeValue(ValueType type, std::complex<T> v) const
+  {
+    return {DimensionalizeValue(type, v.real()), DimensionalizeValue(type, v.imag())};
+  }
```

Comment:
Commenting here along with https://github.com/awslabs/palace/pull/129#discussion_r1392851527, but yes originally I had templated for complex but then decided on this design being nicer. I don't see any penalty for it being templated, so will keep it. If someone is converting an integral type they can add an explicit instantiation when they come upon the linker error.

Reply:
The function doesn't however work for other types because all the constants in `electromagnetics` are double. Adding an explicit instantiation for `int` for instance gives
```
  const auto f1 = iodata.DimensionalizeValue(IoData::ValueType::FREQUENCY, 1.0);
  const auto f2 = iodata.DimensionalizeValue(IoData::ValueType::FREQUENCY, 1);
  std::cout << "f1 " << f1 << " f2 " << f2 << std::endl;
```
prints `f1 0.870683 f2 0` due to the rules on integer multiplication/division and casting. Dropping the template for `double` performs the conversions and gives `f1 0.870683 f2 0.870683` as expected. The extra template syntax buys unexpected behaviour, or requires a user to exercise extra care on numeric types, neither seems ideal.

Reply:
OK maybe it was a bad quick example, you'd never want to instantiate this function for an `int` and if you did then the output you are getting is actually exactly what you'd expect and I argue is actually correct behavior (the returned type should be integral, not floating point in this case). I have a hard time seeing where developers will not be able to take the care to work with the function as it currently exists.